lib/src/tls.rs

```diff
@@ -0,0 +1,1038 @@
+    AddCertificate, CertificateAndKey, CertificateFingerprint, RemoveCertificate,
+    ReplaceCertificate,
+};
+
+use rustls::{
+    internal::pemfile,
+    sign::{CertifiedKey, RSASigningKey},
+    ClientHello, ResolvesServerCert,
+};
+use sha2::{Digest, Sha256};
+use sozu_command::proxy::TlsVersion;
+use x509_parser::{
+    oid_registry::{OID_X509_COMMON_NAME, OID_X509_EXT_SUBJECT_ALT_NAME},
+    pem::{parse_x509_pem, Pem},
+};
+
+// -----------------------------------------------------------------------------
+// CertificateResolver trait
+
+pub trait CertificateResolver {
+    type Error;
+
+    // `get_certificate` return the certificate, its chain and private key
+    fn get_certificate(
+        &self,
+        fingerprint: &CertificateFingerprint,
+    ) -> Option<ParsedCertificateAndKey>;
+
+    // `add_certificate` to the certificate resolver, ensure that it is valid and check if it can
+    // replace another certificates
+    fn add_certificate(
+        &mut self,
+        opts: &AddCertificate,
+    ) -> Result<CertificateFingerprint, Self::Error>;
+
+    // `remove_certificate` from the certificate resolver, may fail if there is no alternative for
+    // a domain name
+    fn remove_certificate(&mut self, opts: &RemoveCertificate) -> Result<(), Self::Error>;
+
+    // `replace_certificate` by a new one, this method is a short-hand for `add_certificate` and
+    // then `remove_certificate`. It is possible that the certificate will not be replaced, if the
+    // new certificate does not match `add_certificate` rules.
+    fn replace_certificate(
+        &mut self,
+        opts: &ReplaceCertificate,
+    ) -> Result<CertificateFingerprint, Self::Error> {
+        self.remove_certificate(&RemoveCertificate {
+            address: opts.address.to_owned(),
+            fingerprint: opts.old_fingerprint.to_owned(),
+        })?;
+
+        let fingerprint = self.add_certificate(&AddCertificate {
+            address: opts.address.to_owned(),
+            certificate: opts.new_certificate.to_owned(),
+            names: opts.new_names.to_owned(),
+            expired_at: opts.new_expired_at.to_owned(),
+        })?;
+
+        Ok(fingerprint)
+    }
+}
+
+// -----------------------------------------------------------------------------
+// CertificateResolverHelper trait
+
+pub trait CertificateResolverHelper {
+    type Error;
+
+    // `find_certificates_by_names` return all fingerprints that are available to provide at least
+    // one name
+    fn find_certificates_by_names(
+        &self,
+        names: &HashSet<String>,
+    ) -> Result<HashSet<CertificateFingerprint>, Self::Error>;
+
+    // `certificate_names` return the hashset of subjects that the certificate is able to handle by
+    // parsing the pem file and scrapping the information
+    fn certificate_names(&self, pem: &Pem) -> Result<HashSet<String>, Self::Error>;
+
+    // `fingerprint` return the computed fingerprint for the given certificate
```

Comment:
please just conjugate the verbs by adding an `s` when the third person is used.

Reply:
Yes, you are right.